Source/Core/Resource.js

```diff
@@ -0,0 +1,534 @@
+         *
+         * @type {String}
+         */
+        this.method = defaultValue(options.method, 'GET');
+
+        /**
+         * Data to be sent with the request.
+         *
+         * @type {Object}
+         */
+        this.data = options.data;
+
+        /**
+         * Overrides the MIME type returned by the server.
+         *
+         * @type {String}
+         */
+        this.overrideMimeType = options.overrideMimeType;
+
+        /**
+         * A proxy to be used when loading the resource.
+         *
+         * @type {DefaultProxy}
+         */
+        this.proxy = options.proxy;
+
+        /**
+         * Function to call when a request for this resource fails. If it returns true, the request will be retried.
+         *
+         * @type {Function}
+         */
+        this.retryCallback = options.retryCallback;
+
+        /**
+         * The number of times the retryCallback should be called before giving up.
+         *
+         * @type {Number}
+         */
+        this.retryAttempts = defaultValue(options.retryAttempts, 0);
+        this._retryCount = 0;
+    }
+
+    /**
+     * A helper function to create a resource depending on whether we have a String or a Resource
+     *
+     * @param {Resource|String} resource A Resource or a String to use when creating a new Resource.
+     * @param {Object{ options If resource is a String, these are the options passed to the Resource constructor. It is ignored otherwise.
+     *
+     * @returns {Resource} If resource is a String, a Resource constructed with the url and options. Otherwise the resource parameter is returned.
+     *
+     * @private
+     */
+    Resource.createIfNeeded = function(resource, options) {
+        if (!defined(resource)) {
+            return;
+        }
+
+        if (typeof resource === 'string') {
+            var args = defined(options) ? clone(options) : {};
+            args.url = resource;
+            resource = new Resource(args);
+        }
+
+        return resource;
+    };
+
+    defineProperties(Resource.prototype, {
+        /**
+         * Query parameters appended to the url.
+         *
+         * @memberof Resource.prototype
+         * @type {Object}
+         *
+         * @readonly
+         */
+        queryParameters: {
+            get: function() {
+                return this._queryParameters;
+            }
+        },
+
+        /**
+         * The key/value pairs used to replace template parameters in the url.
+         *
+         * @memberof Resource.prototype
+         * @type {Object}
+         *
+         * @readonly
+         */
+        templateValues: {
+            get: function() {
+                return this._templateValues;
+            }
+        },
+
+        /**
+         * The url to the resource.
+         *
+         * @memberof Resource.prototype
+         * @type {String}
+         */
+        url: {
+            get: function() {
+                return this.getUrlComponent(true, true);
```

Comment:
Since `getUrlComponent` is trivial, should be cache it?  Or is it too hard to detect when it changes?  I guess it depends on exactly what this vlaue is/

Reply:
A user could get the `queryParameters` or `templateValues` and change them, so it's probably safer to leave as-is.